src/coreclr/src/vm/eventing/eventpipe/ep-rt-coreclr.h

```diff
@@ -0,0 +1,2895 @@
+static
+inline
+bool
+_rt_coreclr_hash_map_is_valid (CONST_HASH_MAP_TYPE *hash_map)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+
+	return (hash_map != NULL && hash_map->table != NULL);
+}
+
+template<typename HASH_MAP_TYPE, typename KEY_TYPE, typename CONST_KEY_TYPE = const KEY_TYPE>
+static
+inline
+void
+_rt_coreclr_hash_map_remove (
+	HASH_MAP_TYPE *hash_map,
+	CONST_KEY_TYPE key)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (hash_map != NULL && hash_map->table != NULL);
+
+	const typename HASH_MAP_TYPE::table_type_t::element_t *ret = NULL;
+	if (hash_map->callbacks.value_free_func)
+		ret = hash_map->table->LookupPtr ((KEY_TYPE)key);
+	hash_map->table->Remove ((KEY_TYPE)key);
+	if (ret)
+		hash_map->callbacks.value_free_func (reinterpret_cast<void *>(static_cast<ptrdiff_t>(ret->Value ())));
+}
+
+template<typename HASH_MAP_TYPE, typename ITERATOR_TYPE, typename CONST_HASH_MAP_TYPE = const HASH_MAP_TYPE>
+static
+inline
+ITERATOR_TYPE
+_rt_coreclr_hash_map_iterator_begin (CONST_HASH_MAP_TYPE *hash_map)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (hash_map != NULL && hash_map->table != NULL);
+
+	return hash_map->table->Begin ();
+}
+
+template<typename HASH_MAP_TYPE, typename ITERATOR_TYPE, typename CONST_HASH_MAP_TYPE = const HASH_MAP_TYPE, typename CONST_ITERATOR_TYPE = const ITERATOR_TYPE>
+static
+inline
+bool
+_rt_coreclr_hash_map_iterator_end (
+	CONST_HASH_MAP_TYPE *hash_map,
+	CONST_ITERATOR_TYPE *iterator)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (hash_map != NULL && hash_map->table != NULL && iterator != NULL);
+
+	return (hash_map->table->End () == *iterator);
+}
+
+template<typename HASH_MAP_TYPE, typename ITERATOR_TYPE>
+static
+inline
+void
+_rt_coreclr_hash_map_iterator_next (ITERATOR_TYPE *iterator)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (iterator != NULL);
+
+	(*iterator)++;
+}
+
+template<typename HASH_MAP_TYPE, typename ITERATOR_TYPE, typename KEY_TYPE, typename CONST_ITERATOR_TYPE = const ITERATOR_TYPE>
+static
+inline
+KEY_TYPE
+_rt_coreclr_hash_map_iterator_key (CONST_ITERATOR_TYPE *iterator)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (iterator != NULL);
+
+	return (*iterator)->Key ();
+}
+
+template<typename HASH_MAP_TYPE, typename ITERATOR_TYPE, typename VALUE_TYPE, typename CONST_ITERATOR_TYPE = const ITERATOR_TYPE>
+static
+inline
+VALUE_TYPE
+_rt_coreclr_hash_map_iterator_value (CONST_ITERATOR_TYPE *iterator)
+{
+	STATIC_CONTRACT_NOTHROW;
+	EP_ASSERT (HASH_MAP_TYPE::table_type_t::s_NoThrow);
+	EP_ASSERT (iterator != NULL);
+
+	return (*iterator)->Value ();
+}
+
+#define EP_RT_DEFINE_LIST_PREFIX(prefix_name, list_name, list_type, item_type) \
```

Comment:
Is it really worth it to have layers of wrapper just to share a trivial growable array implementation? Would it be better for eventpipe to have its own local growable array?

Reply:
For simple array, I agree that we could probably just move away from runtime specifics and do simple implementation directly in library. Maybe we could do that in separate PR since it will affect Mono implementation as well?

Reply:
Just to point out that the wrappers are eliminated at compile time (normally defines + inlined method) so it should be a direct call to underlying runtime specific data structure. For containers there are the following types defined in eventpipe library directly wrapping CoreClr or Mono specific implementations, list, array, queue, hashmap and corresponding iterators for the different container types.

Reply:
Since the wrappers are already implemented for CoreClr and Mono implementing a shim for a new runtime flavour should be straight forward and mainly starting from an existing one replace with that runtime specific implementation, most of the container wrappers are one liners.

Reply:
It assumes that the new flavor has container implementations in C/C++ like that. It is not the case for NativeAOT. It does not have own SArray or equivalent. It does not need them.

Reply:
Ok, what containers do you have available in NativeAOT?

Reply:
Linked list. That's all.

Reply:
OK, so that should solve the list/queue containers, for array's we could potential make simple implementation in eventpipe, what about hashmap?

Reply:
NativeAOT does not have unmanaged implementation of hashmap.

IMHO, a linked list is so simple that it is not worth implementing it using wrappers. The code for the wrappers is more than the implementation itself.

Reply:
The idea was being able to reuse an already existing implementation together with allocator and free callbacks and setup a unified API in eventpipe library using these containers throught the library regardless of how they where implemented (or if implemented in C or C++) on respective runtime. If NativeAOT doesn't have anything that can be re-used, then you will need implementation of these in one way or another.

I guess this is something we could tackle going forward, if library should come with its own set of containers (maybe reuse Mono's containers since they are already in C and battle tested for a long time), but I think we could discuss that outside of this PR, do you agree?

Reply:
Yes, it can be discussed outside of this PR.